File: python/practica/ejercicios_p/figuras.py
import math

class Figuras:
    def __init__(self, base=0, altura=0, radio=0):
        self.base = base
        self.altura = altura
        self.radio = radio

    def bienvenido(self):
        print('Bienvenido a la calculadora de áreas')

    def seleccion(self):
        print('1. Triángulo')
        print('2. Rectángulo')
        print('3. Círculo')
        opcion = int(input('Ingrese la opción: '))
        return opcion

    def ingresar(self):
        if self.nombre == 'Triángulo':
            self.base = float(input('Ingrese la base del triángulo: '))
            self.altura = float(input('Ingrese la altura del triángulo: '))
        elif self.nombre == 'Rectángulo':
            self.base = float(input('Ingrese la base del rectángulo: '))
            self.altura = float(input('Ingrese la altura del rectángulo: '))
        elif self.nombre == 'Círculo':
            self.radio = float(input('Ingrese el radio del círculo: '))

    def _calcular_area(self):
        if self.nombre == 'Triángulo':
            return self.base * self.altura / 2
        elif self.nombre == 'Rectángulo':
            return self.base * self.altura
        elif self.nombre == 'Círculo':
            return math.pi * self.radio ** 2

    def mostrar(self):
        area = self._calcular_area()
        if self.nombre == 'Triángulo':
            print(f'El área del triángulo es: {area}')
        elif self.nombre == 'Rectángulo':
            print(f'El área del rectángulo es: {area}')
        elif self.nombre == 'Círculo':
            print(f'El área del círculo es: {area}')
        else:
            print('Opción no válida')

class Triangulo(Figuras):
    def __init__(self, base=0, altura=0):
        super().__init__(base, altura)
        self.nombre = 'Triángulo'

class Rectangulo(Figuras):
    def __init__(self, base=0, altura=0):
        super().__init__(base, altura)
        self.nombre = 'Rectángulo'

class Circulo(Figuras):
    def __init__(self, radio=0):
        super().__init__(radio=radio)
        self.nombre = 'Círculo'

def main():
    fig = Figuras()
    fig.bienvenido()
    opcion = fig.seleccion()
    if opcion == 1:
        figura = Triangulo()

File: js/POO/claseAbtracta/figuras.js
"use strict";
const readline = require("readline");

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

class Forma {
    constructor() {
        if (new.target === Forma) {
            throw new Error("No se puede instanciar la clase abstracta Forma");
        }
    }

    calcularArea() {
        throw new Error("El método calcularArea debe ser implementado en la clase derivada");
    }

    calcularPerimetro() {
        throw new Error("El método calcularPerimetro debe ser implementado en la clase derivada");
    }

    mostrar() {
        console.log(`Información sobre la forma: ${this.constructor.name}`);
        console.log(`Área: ${this.calcularArea()}`);
        console.log(`Perímetro: ${this.calcularPerimetro()}`);
    }

    ingresar(guardar) {
        throw new Error("Método ingresar debe ser implementado en la clase derivada");
    }

    seleccionarOperacion(guardar) {
        throw new Error("Método seleccionarOperacion debe ser implementado en la clase derivada");
    }
}

class Rectangulo extends Forma {
    constructor(_alto = 0, _ancho = 0) {
        super();
        this._alto = _alto;
        this._ancho = _ancho;
    }

    calcularArea() {
        return this._alto * this._ancho;
    }

    calcularPerimetro() {
        return 2 * (this._alto + this._ancho);
    }

    ingresar(guardar) {
        rl.question("Ingrese el alto del rectángulo: ", (alto) => {
            this._alto = parseFloat(alto);
            rl.question("Ingrese el ancho del rectángulo: ", (ancho) => {
                this._ancho = parseFloat(ancho);
                guardar();
            });
        });
    }

    seleccionarOperacion(guardar) {
        console.log("\n¿Qué deseas hacer con el rectángulo?");
        console.log("1. Calcular área");
        console.log("2. Calcular perímetro");
        console.log("3. Mostrar información");
        console.log("4. Salir");

        rl.question("Ingrese una opción: ", (opcion) => {
            switch (opcion) {
                case '1':
                    console.log(`Área: ${this.calcularArea()}`);
                    this.seleccionarOperacion(guardar);
                    break;
                case '2':
                    console.log(`Perímetro: ${this.calcularPerimetro()}`);
                    this.seleccionarOperacion(guardar);
                    break;
                case '3':
                    this.mostrar();
                    this.seleccionarOperacion(guardar);
                    break;
                case '4':
                    rl.close();
                    break;
                default:
                    console.log("Opción no válida. Intente nuevamente.");
                    this.seleccionarOperacion(guardar);
                    break;
            }
        });
    }
}

class Circulo extends Forma {
    constructor(radio = 0) {
        super();
        this.radio = radio;
    }

    calcularArea() {
        return Math.PI * Math.pow(this.radio, 2);
    }

    calcularPerimetro() {
        return 2 * Math.PI * this.radio;
    }

    ingresar(guardar) {
        rl.question("Ingrese el radio del círculo: ", (radio) => {
            this.radio = parseFloat(radio);
            guardar();
        });
    }

    seleccionarOperacion(guardar) {
        console.log("\n¿Qué deseas hacer con el círculo?");
        console.log("1. Calcular área");
        console.log("2. Calcular perímetro");
        console.log("3. Mostrar información");
        console.log("4. Salir");

        rl.question("Ingrese una opción: ", (opcion) => {
            switch (opcion) {
                case '1':
                    console.log(`Área: ${this.calcularArea()}`);
                    this.seleccionarOperacion(guardar);
                    break;
                case '2':
                    console.log(`Perímetro: ${this.calcularPerimetro()}`);
                    this.seleccionarOperacion(guardar);
                    break;
                case '3':
                    this.mostrar();
                    this.seleccionarOperacion(guardar);
                    break;
                case '4':
                    rl.close();
                    break;
                default:
                    console.log("Opción no válida. Intente nuevamente.");
                    this.seleccionarOperacion(guardar);
                    break;
            }
        });
    }
}

// Este será el flujo general, sin necesidad de lógica dentro del main.
function seleccionarForma(guardar) {
    console.log("Selecciona una forma para trabajar:");
    console.log("1. Rectángulo");
    console.log("2. Círculo");
    console.log("3. Salir");

    rl.question("Ingrese una opción: ", (opcion) => {
        switch (opcion) {
            case '1':
                const rectangulo = new Rectangulo();
                rectangulo.ingresar(() => {
                    rectangulo.seleccionarOperacion(guardar);
                });
                break;
            case '2':
                const circulo = new Circulo();
                circulo.ingresar(() => {
                    circulo.seleccionarOperacion(guardar);
                });
                break;
            case '3':
                rl.close();
                break;
            default:
                console.log("Opción no válida. Intente nuevamente.");
                seleccionarForma(guardar);
                break;
        }
    });
}

function main() {
    seleccionarForma(() => {});
}

main();
